Dynamic quantizatiion for Fx

DeepDriving opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49948 Dynamic quantizatiion for Fx**
* #49266 compare_model_outputs_fx API implementation
* #48951 compare_model_stub_fx API implementation

Dynamic quantization for FX
buck test @mode/dev caffe2/test:quantization_fx -- 'test_compare_weights_lstm_dynamic_fx'
buck test @mode/dev caffe2/test:quantization_fx -- 'test_compare_model_stub_linear_dynamic_fx'
buck test @mode/dev caffe2/test:quantization_fx -- 'test_compare_model_stub_lstm_dynamic_fx'
buck test @mode/dev caffe2/test:quantization_fx -- 'test_compare_model_outputs_linear_dynamic_fx'

Differential Revision: [D25726967](https://our.internmc.facebook.com/intern/diff/D25726967/)

Commit:
Dynamic quantizatiion for Fx

Dynamic quantization for FX
buck test @mode/dev caffe2/test:quantization_fx -- 'test_compare_weights_lstm_dynamic_fx'
buck test @mode/dev caffe2/test:quantization_fx -- 'test_compare_model_stub_linear_dynamic_fx'
buck test @mode/dev caffe2/test:quantization_fx -- 'test_compare_model_stub_lstm_dynamic_fx'
buck test @mode/dev caffe2/test:quantization_fx -- 'test_compare_model_outputs_linear_dynamic_fx'

Differential Revision: [D25726967](https://our.internmc.facebook.com/intern/diff/D25726967/)

[ghstack-poisoned]

hx89 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -285,3 +382,32 @@ def compare_and_validate_results(float_model, q_model, data):
+        def compare_and_validate_results(float_model, q_model, data):
+            act_compare_dict = compare_model_outputs_fx(float_model, q_model, data)
+            print(act_compare_dict)
+            expected_act_compare_dict_keys = {"stats"}
```
Expected key should have another one for fc1.

DeepDriving replied:
Fixed.

Commit:
Update on "Dynamic quantizatiion for Fx"


Dynamic quantization for FX
buck test @mode/dev caffe2/test:quantization_fx -- 'test_compare_weights_lstm_dynamic_fx'
buck test @mode/dev caffe2/test:quantization_fx -- 'test_compare_model_stub_linear_dynamic_fx'
buck test @mode/dev caffe2/test:quantization_fx -- 'test_compare_model_stub_lstm_dynamic_fx'
buck test @mode/dev caffe2/test:quantization_fx -- 'test_compare_model_outputs_linear_dynamic_fx'

Differential Revision: [D25726967](https://our.internmc.facebook.com/intern/diff/D25726967/)

[ghstack-poisoned]

hx89 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -135,6 +136,34 @@ def compare_and_validate_results(float_model, q_model):
+            weight_dict = compare_weights_fx(
+                float_model.state_dict(), q_model.state_dict()
+            )
+            self.assertEqual(len(weight_dict), 1)
```
Could you add the check if the key in weight_dict is expected?

DeepDriving replied:
Added in compare_model_stub PR

hx89 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -214,6 +243,77 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
+            ob_dict = compare_model_stub_fx(
+                float_model, q_model, module_swap_list, data
+            )
+            self.assertEqual(len(ob_dict), 1)
```
Could you add the check if the key in ob_dict is expected?

hx89 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -214,6 +243,77 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
+            ob_dict = compare_model_stub_fx(
+                float_model, q_model, module_swap_list, input, hidden
+            )
+            self.assertEqual(len(ob_dict), 1)
```
Could you add the check if the key in ob_dict is expected?

DeepDriving replied:
Added in compare_model_stub PR

Commit:
Update on "Dynamic quantizatiion for Fx"


Dynamic quantization for FX
buck test @mode/dev caffe2/test:quantization_fx -- 'test_compare_weights_lstm_dynamic_fx'
buck test @mode/dev caffe2/test:quantization_fx -- 'test_compare_model_stub_linear_dynamic_fx'
buck test @mode/dev caffe2/test:quantization_fx -- 'test_compare_model_stub_lstm_dynamic_fx'
buck test @mode/dev caffe2/test:quantization_fx -- 'test_compare_model_outputs_linear_dynamic_fx'

Differential Revision: [D25726967](https://our.internmc.facebook.com/intern/diff/D25726967/)

[ghstack-poisoned]

raghuramank100 reviewed (commented):
Is the change in this PR mainly the additional tests?
If so, the changes to _numeric_suite_fx dont seem to be used in the code.
